In [11]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import  StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
import numpy as np
import pandas as pd

data =pd.read_csv("Clean_data.csv")


In [13]:
data.columns

Index(['category', 'num_bedrooms', 'num_bathrooms', 'surface_area',
       'transaction_type', 'price', 'city', 'region', 'log_price'],
      dtype='object')

In [15]:
import pandas as pd


# One-hot encode the categorical columns
c1 = pd.get_dummies(data['category'])
c2 = pd.get_dummies(data['transaction_type'])
c3 = pd.get_dummies(data['city'])
c4 = pd.get_dummies(data['region'])

# Concatenate the original DataFrame with the new one-hot encoded columns
merged = pd.concat([data, c1, c2, c3, c4], axis='columns')

# Drop the original categorical columns
final = merged.drop(['category', 'transaction_type', 'city', 'region'], axis='columns')

# Display the final DataFrame
final


num_bedrooms  num_bathrooms  surface_area      price  log_price  \
0              -1.0           -1.0          -1.0   100000.0   5.000000   
1              -1.0           -1.0          -1.0   316000.0   5.499687   
2               2.0            1.0          80.0      380.0   2.579784   
3               1.0            1.0          90.0       70.0   1.845098   
4               2.0            2.0         113.0   170000.0   5.230449   
...             ...            ...           ...        ...        ...   
12581          -1.0           -1.0          -1.0  3200000.0   6.505150   
12582           1.0            1.0         100.0      600.0   2.778151   
12583           3.0            1.0         760.0  1950000.0   6.290035   
12584           3.0            1.0         190.0   240000.0   5.380211   
12585           2.0            1.0          70.0      500.0   2.698970   

       Appartements  Bureaux et Plateaux  Colocations  Locations de vacances  \
0             False                False        False                  False   
1             False                False        False                  False   
2              True                False        False                  False   
3             False                False        False                   True   
4              True                False        False                  False   
...             ...                  ...          ...                    ...   
12581         False                False        False                  False   
12582          True                False        False                  False   
12583         False                False        False                  False   
12584         False                False        False                  False   
12585         False                False        False                  False   

       Magasins, Commerces et Locaux industriels  ...  Tozeur  Tunis  \
0                                          False  ...   False  False   
1                                          False  ...   False  False   
2                                          False  ...   False  False   
3                                          False  ...   False  False   
4                                          False  ...   False  False   
...                                          ...  ...     ...    ...   
12581                                      False  ...   False  False   
12582                                      False  ...   False  False   
12583                                      False  ...   False  False   
12584                                      False  ...   False  False   
12585                                      False  ...   False  False   

       Téboulba  Téboursouk  Utique  Zaghouan  Zaouit-Ksibat Thrayett  Zarzis  \
0         False       False   False     False                   False   False   
1         False       False   False     False                   False   False   
2         False       False   False     False                   False   False   
3         False       False   False     False                   False   False   
4         False       False   False     False                   False   False   
...         ...         ...     ...       ...                     ...     ...   
12581     False       False   False     False                   False   False   
12582     False       False   False     False                   False   False   
12583     False       False   False     False                   False   False   
12584     False       False   False     False                   False   False   
12585     False       False   False     False                   False   False   

       Zarzouna  Zéramdine  
0         False      False  
1         False      False  
2         False      False  
3         False      False  
4         False      False  
...         ...        ...  
12581     False      False  
12582     False      False  
12583     False      False  
12584     False      False  


In [17]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

X = final.drop(columns=['price', 'log_price'])  # Remove price and log_price from features
y = final['log_price']  # Using log_price as the target variable

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#  Feature scaling (important for regression models)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [28]:
from sklearn.linear_model import LinearRegression, Lasso
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score


models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(n_estimators=100, random_state=42),
    
}

# Iterate through the models, train, and evaluate
for model_name, model in models.items():

    model.fit(X_train_scaled, y_train)

    predictions = model.predict(X_test_scaled)

    # Evaluate the model
    print(f"\n{model_name} Performance:")
    print(f"Mean Squared Error (MSE): {mean_squared_error(y_test, predictions):,.2f}")
    print(f"Mean Absolute Error (MAE): {mean_absolute_error(y_test, predictions):,.2f}")
    print(f"R^2 Score: {r2_score(y_test, predictions):,.4f}")
    print("model score :", model.score(X_train_scaled, y_train)) # the R² score of the model on the training data.

    cv = ShuffleSplit(n_splits=5, test_size=0.2, random_state=0)
    print(cross_val_score(model, X, y, cv=cv))

#  If you want to convert predictions back to the original price scale
#y_pred_price = 10 ** y_pred 



Linear Regression Performance:
Mean Squared Error (MSE): 18,161,291,311,517,788,012,544.00
Mean Absolute Error (MAE): 8,076,319,991.67
R^2 Score: -9,553,821,214,252,407,455,744.0000
0.8339215366260915
[ 8.13313445e-01 -4.16319691e+04  8.22054091e-01  8.01110838e-01
  7.78609224e-01]

Decision Tree Performance:
Mean Squared Error (MSE): 0.42
Mean Absolute Error (MAE): 0.36
R^2 Score: 0.7783
0.9366226117553299
[0.78373803 0.77176119 0.7819205  0.76890133 0.80763427]

Random Forest Performance:
Mean Squared Error (MSE): 0.31
Mean Absolute Error (MAE): 0.31
R^2 Score: 0.8368
0.9239752795097187
[0.83620139 0.80890007 0.8298679  0.81116909 0.85217551]

Gradient Boosting Performance:
Mean Squared Error (MSE): 0.30
Mean Absolute Error (MAE): 0.32
R^2 Score: 0.8433
0.8507399756850565
[0.83681617 0.81321654 0.83822727 0.82136832 0.86103539]


| **Model**           | **Mean Squared Error (MSE)**       | **Mean Absolute Error (MAE)**       | **R² Score**               |
|---------------------|------------------------------------|-------------------------------------|----------------------------|
| **Linear Regression**| 18,161,291,311,517,788,012,544.00  | 8,076,319,991.67                   | -9,553,821,214,252,407,455,744.0000 |
| **Decision Tree**    | 0.42                               | 0.36                                | 0.7783                     |
| **Random Forest**    | 0.31                               | 0.31                                | 0.8368                     |
| **Gradient Boosting**| 0.30                               | 0.32                                | 0.8433                     |

### Summary:
- **Best Performer**: **Gradient Boosting** (MSE: 0.30, MAE: 0.32, R²: 0.8433)
- **Second Best**: **Random Forest** (MSE: 0.31, MAE: 0.31, R²: 0.8368)
- **Third Best**: **Decision Tree** (MSE: 0.42, MAE: 0.36, R²: 0.7783)
- **Worst Performer**: **Linear Regression** (MSE: 18,161,291,311,517,788,012,544.00, MAE: 8,076,319,991.67, R²: -9,553,821,214,252,407,455,744.0000)
